### PR TITLE
Stub for Delta Eddington solver

### LINKED ISSUE #65: Create stub class for Delta Eddington solver in C++
Create a stub class in C++ for the new Delta Eddington solver, and add an extending class of `solver_t` in fortran that will call the C++ solver functions to solve the radiation field. The solver class is defined here: https://github.com/NCAR/tuv-x/blob/d60e6c037ee0c481d629d5b9bf977f89a29d1b35/src/radiative_transfer/solver.F90#L33-L42

**Acceptance Criteria**
- A stub class for the Delta Eddington solver in C++ is included in the build.
- Users can select the C++ Delta Eddington solver as a configuration option.
- If selected the Delta Eddington solver functions will be called through a Fortran wrapper that extends `solver_t` when the radiation field is calculated and just returns all zeros.

**Ideas**
- The solver can be specified in the configuration here: https://github.com/NCAR/tuv-x/blob/d60e6c037ee0c481d629d5b9bf977f89a29d1b35/examples/tuv_5_4.json#L102-L104
-

### LINKED ISSUE #67: Create grid class in C++
Create a C++ grid class or struct that provide the same functionality as the Fortran `grid_t` class: https://github.com/NCAR/tuv-x/blob/d60e6c037ee0c481d629d5b9bf977f89a29d1b35/src/grid.F90#L15-L35

(without the MPI functions)

**Acceptance Criteria**
- Grid instances can be created and can provide mid-point, interface, and cell width (difference between adjacent interfaces) values for each grid section.

**Ideas**
- Consider just having the mid-point and interface/edge values save as data members, and calculated the delta/width values as needed.
- The `handle_` was primarily used to lookup grids by name. Use `std::map` instead to manage collections of grids.

### LINKED ISSUE #68: Create profile class in C++
Create a C++ profile class that provides the same functionality as the Fortran `profile_t` class: https://github.com/NCAR/tuv-x/blob/d60e6c037ee0c481d629d5b9bf977f89a29d1b35/src/profile.F90#L15-L40

(without the MPI functions)

**Acceptance Criteria**
- Instances of profiles can be created and populated with values at mid-points and interfaces.

**Ideas**
- Only have the mid-point and interface arrays as data members, calculate the layer densities as needed (burden densities should not be needed for now)
- The `handle_` was primarily used to lookup profiles by name. Use `std::map` instead to manage collections of profiles.

### PR DESCRIPTION
**Pull request by mattldawson**
Adds a stub class for the C++/CUDA Delta Eddington solver.

The code is structured in a way that, hopefully, CUDA versions of the `Array2D` and `Array3D` classes will be able to be used similarly to how the `CudaDenseMatrix` and `CudaSparseMatrix` are used in MICM. Users would handle loading input array data, synching to device, running TUV-x, and synching output data back to host.

I used policies to share as much code as possible between CPU and GPU versions of the solver. @K20shores - if there is a better way to structure this with `auto`, I'm happy to rework this, but it wasn't obvious to me how to do this.

closes #65, #67, #68

### CODE REVIEW

**Review by K20shores — APPROVED** [review]
love it! especially since the asymmetry parameter is fully spelled out instead of just `g`

**Review by sjsprecious — APPROVED** [review]
Thanks @mattldawson for initiating this effort. Very nice work!

File: include/tuvx/util/array3d.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  const T &operator()(size_t i, size_t j, size_t k) const {
+    return data_[index(i, j, k)];
+  }
+
+  size_t size1() const { return dim1_; }
+
+  size_t size2() const { return dim2_; }
+
+  size_t size3() const { return dim3_; }
+
```
Comment (boulderdaze) [inline include/tuvx/util/array3d.hpp]: In`array2d.hpp`, the `size()` functions are in CamelCase.  Would it be possible to make them in the same style for consistency?

**Review by boulderdaze — CHANGES_REQUESTED** [review]
Looks good! Just have a single hopefully minor change request.

File: include/tuvx/util/array3d.hpp (hunk 1 above)
Comment (mattldawson) [reply include/tuvx/util/array3d.hpp]: done!

**Review by boulderdaze — APPROVED** [review]
Thank you!

File: include/tuvx/radiative_transfer/solvers/delta_eddington.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    DeltaEddington() = default;
+
+    /// @brief Solve the radiative transfer equation for a collection of columns
+    /// @param solar_zenith_angles Solar zenith angles for each column [radians].
+    /// @param grids Grids available for the radiative transfer calculation.
+    /// @param profiles Profiles available for the radiative transfer calculation.
+    /// @param radiation_field The calculated radiation field.
+    ///
+    /// Solves two-stream equations for multiple layers. These routines are based
+    /// on equations from: Toon et al., J.Geophys.Res., v94 (D13), Nov 20, 1989.
```
Comment (mwaxmonsky) [inline include/tuvx/radiative_transfer/solvers/delta_eddington.hpp:32]: Could we add the DOI?
https://doi.org/10.1029/JD094iD13p16287

File: include/tuvx/radiative_transfer/solvers/delta_eddington.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    /// It contains 9 two-stream methods to choose from. A pseudo-spherical
+    /// correction has also been added.
+    ///
+    /// The original delta-Eddington paper is:
+    /// Joseph and Wiscombe, J. Atmos. Sci., 33, 2453-2459, 1976
```
Comment (mwaxmonsky) [inline include/tuvx/radiative_transfer/solvers/delta_eddington.hpp:38]: Similarily:
https://doi.org/10.1175/1520-0469(1976)033%3C2452:TDEAFR%3E2.0.CO;2

**Review by mwaxmonsky — APPROVED** [review]
Looks great @mattldawson!  Just had two minor recommendations if they're correct but definitely not worth holding up the PR.

File: include/tuvx/radiative_transfer/solvers/delta_eddington.hpp (hunk 2 above)
Comment (mattldawson) [reply include/tuvx/radiative_transfer/solvers/delta_eddington.hpp:32]: done!

File: include/tuvx/radiative_transfer/solvers/delta_eddington.hpp (hunk 3 above)
Comment (mattldawson) [reply include/tuvx/radiative_transfer/solvers/delta_eddington.hpp:38]: done!

**Pull request merged**